### PR TITLE
Use faster hash for short arrays.

### PR DESCRIPTION
**Pull request by oscardssmith**
Since the current code already hashes all elements of small arrays (<4096 elements), this is basically a fast-path that avoids a lot of the math. It also xors the hashes, which should allow the compiler to vectorize hashing. As a result, I'm measuring about a 4x speedup for hashing a 20x20 matrix, and have not found cases where this is slower. The biggest thing this needs is someone more familiar with `Array` hashing to make sure I'm not missing anything obvious about how hashing should work.

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -2403,7 +2403,14 @@ function hash(A::AbstractArray, h::UInt)
     # Instead hash the tuple of firsts and lasts along each dimension
     h = hash(map(first, axes(A)), h)
     h = hash(map(last, axes(A)), h)
-    isempty(A) && return h
+
+    # For short arrays, it's not worth doing anything complicated 
+    if length(A) < 10000
```
Comment (mbauman) [inline base/abstractarray.jl:2408]: Any rough tests on this threshold?  Seems like we'd want it to jive with the 4096 below.

Comment (oscardssmith) [reply base/abstractarray.jl:2408]: For a `10000` length `Vector{Float64}`, I'm seeing a 7% speedup (more for matrices of the same size, not sure why).  The reason this can still be beneficial is that the logic is enough simpler that it can still be faster while doing 50% more hashes. That said, the tradeoff will be less good for `Array`s of more complex objects. Given that, I could see arguments both ways. What are your thoughts?

File: base/abstractarray.jl [hunk 2]
```diff
@@ -2403,7 +2403,14 @@ function hash(A::AbstractArray, h::UInt)
...
+        @inbounds for p in pairs(IndexLinear(),A)
```
Comment (mbauman) [inline base/abstractarray.jl:2409]: Do we really need to hash the key & value?  Below, too.  This might be an artifact of how we used to do it (which _needed_ to hash the key since we did value-based skipping).  We already hash the axes, and the access pattern will be constant for a given shape.

File: base/abstractarray.jl [hunk 3]
```diff
@@ -2403,7 +2403,14 @@ function hash(A::AbstractArray, h::UInt)
...
+            h ⊻= hash(p)
```
Comment (mbauman) [inline base/abstractarray.jl:2410]: Although if we don't include the key, this will probably need to be `h = hash(p, h)` to prevent evenly repeated values from hashing the same.

Comment (oscardssmith) [reply base/abstractarray.jl:2410]: Yeah. The reason I wanted to do it this way was so the `hash` computations would be independent. This theoretically should allow the compiler to re-order this to interweave hash computation.

Comment (mbauman) [reply base/abstractarray.jl:2410]: Is it cheaper to do half the hashes but use the `h = hash(x, h)` form? We could/should remove the key hashing below, too.

Comment (oscardssmith) [reply base/abstractarray.jl]: Good call. Updated

File: base/abstractarray.jl [hunk 4]
```diff
@@ -2434,9 +2441,8 @@ function hash(A::AbstractArray, h::UInt)
     n = 0
     while true
```
Comment (mbauman) [inline base/abstractarray.jl:2442]: Why not also:
```suggestion
    @inbounds while true
```

I vaguely remember skipping `@inbounds` when we first implemented this because it felt fancy... but we've never gotten out of bounds reports.

**Comment by oscardssmith** [comment]
Any idea what's wrong with CI?

File: base/abstractarray.jl [hunk 5]
```diff
@@ -2403,7 +2403,14 @@ function hash(A::AbstractArray, h::UInt)
     # Instead hash the tuple of firsts and lasts along each dimension
     h = hash(map(first, axes(A)), h)
     h = hash(map(last, axes(A)), h)
-    isempty(A) && return h
+
+    # For short arrays, it's not worth doing anything complicated
+    if length(A) < 8192
+        @inbounds for x in A
```
Comment (vtjnash) [inline base/abstractarray.jl:2409]:
```suggestion
        for x in A
```
This annotation seems like nonsense here?

(style nit: there should usually also be a space after a comma, on the next line)

File: base/abstractarray.jl [hunk 6]
```diff
@@ -2432,11 +2439,10 @@ function hash(A::AbstractArray, h::UInt)
     fibskip = prevfibskip = oneunit(linidx)
     first_linear = first(LinearIndices(linear_to_key))
     n = 0
-    while true
+    @inbounds while true
```
Comment (vtjnash) [inline base/abstractarray.jl:2442]: This seems unlikely to matter, would be an unrelated change, and we typically consider this annotation to be undesirable unless strongly proven otherwise.
```suggestion
    while true
```

**Comment by oscardssmith** [comment]
Only test failure was a timeout. Good to merge?

**Comment by oscardssmith** [comment]
Merging in 24 hours unless anyone has objections.

File: base/abstractarray.jl [hunk 7]
```diff
@@ -2434,9 +2441,8 @@ function hash(A::AbstractArray, h::UInt)
     n = 0
     while true
         n += 1
-        # Hash the current key-index and its element
-        elt = A[keyidx]
```
Comment (vtjnash) [inline base/abstractarray.jl:2438]: PkgEval notes that `elt` is used later in this function and is now causing UndefVarError in some packages last night. See for example:

https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_date/2021-03/17/DataStructures.1.7.0-DEV-d7da2a4a32.log

Comment (oscardssmith) [reply base/abstractarray.jl:2438]: oh. Good catch. Will make a new PR to fix this.
